### PR TITLE
New commands to fetch key details

### PR DESCRIPTION
**Pull request by uddhav-dave**
### New Commands
- `sq key info` is a command for fetching details on key by its name (provided with `--dsm-key` arg). If keys with same name exists, it prints details on all of them.
- `sq key list-dsm-keys` is a commands for listing keys the app has access over. Verbosity on details of each key can be controlled by a feature flag `--long`. This command will only print details on PGP keys (i.e, keys with PGP metadata)

Features are added using readily available APIs

### CODE REVIEW

File: openpgp-dsm/src/lib.rs [hunk 1]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
...
+    }
+
+    let footer = format!("\nTOTAL OBJECTS: {}\n", key_info_store.len());
+    key_info_store.push(footer);
+
+    Ok(key_info_store)
+}
+/// Iterates through accessible groups and fetches all keys available to app.
+/// Returns an sorted list of keys grouped by group ID.
+/// Note that this funtion has linear time complexity over the number of groups
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:848]: Nit: I would drop the last sentence, as this function could be threaded in the future.

File: openpgp-dsm/src/lib.rs [hunk 2]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
...
+        key.kid.unwrap_or(Uuid::nil()),
+        key.created_at.to_datetime(),
+        20,
+        name = key.name.as_ref().unwrap_or(&"nil".to_owned()),
+    )
+}
+
+/// Prints key details in verbose format, includes all fields.
+/// Expects `DSM_LABEL_PGP` key to be present in metadata.
+/// !!! PANICS if metadata is corrupted or KeyID not present
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:909]: With the `DsmKeyStruct`, no need to panic here, but rather error before (when parsing the data into the struct)

File: sq/src/sq-usage.rs [hunk 3]
```diff
@@ -768,6 +770,67 @@
 //! $ sq key attest-certifications --none juliet.pgp
 //! ```
 //!
+//! ### Subcommand key info
+//!
+//! ```text
+//!
+//! This command will print data on the given DSM key name if the key is present.
```
Comment (zugzwang) [inline sq/src/sq-usage.rs:777]: Nit: use present tense as other commands: will print -> prints

File: openpgp-dsm/src/lib.rs [hunk 4]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
...
+    }
+
+    let footer = format!("\nTOTAL OBJECTS: {}\n", key_info_store.len() - 1);
+    key_info_store.push(footer);
+
+    Ok(key_info_store)
+}
+
+/// Filter keys which do not have PGP signature/metadata
+fn filter_pgp_keys(key: &Sobject) -> bool {
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:889]: This function is only used in `list_keys` and is 4 lines, so it should be defined inside it.

File: openpgp-dsm/src/lib.rs [hunk 5]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
...
+    match &key.custom_metadata {
+        Some(metadata) => metadata.contains_key(&DSM_LABEL_PGP.to_string()),
+        None => false,
+    }
+}
+
+/// Prints key details in concise format, includes name, uuid, created_at
+fn sobject_short_details_formatter(key: &Sobject) -> String {
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:897]: I would define a `DsmKeyInfo` struct, and have this function as a member, same for the `long` function.

**Review by zugzwang — COMMENTED** [review]
Please write a test script that generates one ore more keys, also imports one or more keys, and then list those?

Could you also rebase and bump the version?

File: openpgp-dsm/src/lib.rs [hunk 6]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
     Ok(())
 }
 
+/// Gets info on a key and prints revelant PGP details for it.
+/// Returns `Err` if key is not present.
+pub fn dsm_key_info(cred: Credentials, key_name: &str) -> Result<Vec<String>> {
```
[the remark this answers is not in the document]
Comment (uddhav-dave) [reply openpgp-dsm/src/lib.rs:824]: okay, ill add a struct to this.
about the `Result<Vec<_>>` , yes we don't need `Vec` here. I verified no 2 keys can have the same name regardless of group.

File: openpgp-dsm/src/lib.rs [hunk 7]
```diff
@@ -818,6 +819,116 @@ pub fn generate_key(
...
+    let mut key_info_store: Vec<String> = Vec::new();
+
+    let params = ListSobjectsParams {
+        name: Some(key_name.to_string()),
+        ..Default::default()
+    };
+
+    let keys = dsm_client.list_sobjects(Some(&params))?;
+
+    // If multiple keys with same name exists, iterate and print them all
```
[the remark this answers is not in the document]
Comment (uddhav-dave) [reply openpgp-dsm/src/lib.rs:836]: cannot happen, will rectify

File: openpgp-dsm/src/lib.rs [hunk 8]
```diff
@@ -885,69 +885,112 @@ pub fn generate_key(
...
+    hashing_algo: Option<HashAlgorithm>,
+    created_at: SdkmsTime,
+    last_used_at: SdkmsTime,
+    fingerprint: String,
+}
+
+impl From<&Sobject> for DsmKeyInfo {
+    /// Expects `DSM_LABEL_PGP` key to be present in metadata.
+    fn from(key: &Sobject) -> Self {
+        let key_md = KeyMetadata::from_sobject(&key).unwrap();
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:900]: Could we use `TryFrom` instead? In order to give a nice error to the user.

File: openpgp-dsm/src/lib.rs [hunk 9]
```diff
@@ -885,69 +885,112 @@ pub fn generate_key(
...
+            created_at: key.created_at,
+            last_used_at: key.lastused_at,
+            fingerprint: key_md.fingerprint
+        }
+    }
+}
+
+impl DsmKeyInfo {
+    /// Prints key details in concise format, includes name, uuid, created_at
+     pub fn sobject_short_details_formatter(&self) -> String {
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:914]: Nit: When I first read the function name, I thought a formatter was to be returned.

It's a bit odd to have `let s: String = key_info.sobject_short_details_formatter();`. I think it is more readable to have `let `s: String = key_info.format_short();` or similar.

File: openpgp-dsm/src/lib.rs [hunk 10]
```diff
@@ -885,69 +885,112 @@ pub fn generate_key(
...
-    // If multiple keys with same name exists, iterate and print them all
-    for key in keys.iter() {
-        key_info_store.push(sobject_long_details_formatter(&key));
-    }
-
-    let footer = format!("\nTOTAL OBJECTS: {}\n", key_info_store.len());
-    key_info_store.push(footer);
+    let key: DsmKeyInfo = match dsm_client.list_sobjects(Some(&params))?.first() {
+        Some(key) => key.into(),
+        None => return Ok(None)
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:964]: A bit curious; when do we return `Ok(None)` here?

**Review by zugzwang — APPROVED** [review]
Last nits. Otherwise approved, thanks for the PR.

File: openpgp-dsm/src/lib.rs [hunk 11]
```diff
@@ -884,6 +885,119 @@ pub fn generate_key(
     Ok(())
 }
 
+pub struct DsmKeyInfo {
+    name: String,
+    kid: Uuid,
+    hashing_algo: Option<HashAlgorithm>,
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:891]: Hashing algorithm? Why?
We should be showing the crypto algorithm of the primary key. E.g. RSA (bitsize) or EC (curve)

File: sq/tests/dsm/print_dsm_key_info.sh [hunk 12]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash -e
```
Comment (zugzwang) [inline sq/tests/dsm/print_dsm_key_info.sh:1]: Add this script to the Makefile so it gets run with `make test-dsm`

File: openpgp-dsm/src/lib.rs (hunk 11 above)
Comment (zugzwang) [reply openpgp-dsm/src/lib.rs:891]: When I run the test script I get:
```
generate-knownkeys-test-IMW0VcxkpY-cv25519:
    UUID: c0f31763-5c4f-45d9-85ab-6d05aeb58c45
    Hashing algorithm: None
    Created at: 2023-02-20 21:11:10 UTC
    Last used at: NA
    PGP fingerprint: 17D0999363E48109BE2003CC2E864D5120879189
```

**Pull request merged**